include/cxxeffect/task.hpp

```diff
@@ -0,0 +1,53 @@
+#pragma once
+#include <functional>
+#include <future>
+
+namespace eff {
+    template <class F, class T>
+    using flatmap_t = decltype(std::declval<F>()(std::declval<T>()).unsafeRunSync());
+    template <class F, class T>
+    using map_t = decltype(std::declval<F>()(std::declval<T>()));
+    template <class T>
+    class task {
+        std::function<T()> thunk;
+       public:
+        task() = default;
+        task(task const&) = default;
+        task(task&&) = default;
+        task(std::function<T()> thunk) : thunk(std::move(thunk)) {}
+
+        static task pure(T value) {
+            return task([=]() { return value; });
+        }
+        static auto delay(std::function<T()> thunk) {
+            return task(thunk);
+        }
+
+        template <class F>
+        auto map(F func) const -> task<map_t<F, T>> {
+            return {
+                    [f = std::move(func), input = *this]() {
+                    return f(input.unsafeRunSync());
+                }
+            };
+        }
+
+        template <class F>
+        auto flatMap(F func) const -> task<flatmap_t<F, T>> {
+            return {
+                [f = func, ta = *this]() {
+                    return f(ta.unsafeRunSync()).unsafeRunSync();
+                }
+            };
+        }
+
+        T unsafeRunSync() const {
+            return thunk();
```

Comment:
I don't think this is stack-safe, is it? Won't you be diving into a bunch of `thunk()` calls?

Reply:
Alternative 1: That doesn't solve the issue with `flat_map` does it?
Alternative 2: If we're using Coroutines anyway, we're doing a context switch/yield on async boundaries... I _think_ what cats-effect does is it also just introduces a yield when the stack gets too deep.
Alternative 3: I'm pretty sure there's a way to implement `flat_map` entirely with tail-recursion. `task` with a thunk is a lot more like [Eval](https://github.com/typelevel/cats/blob/main/core/src/main/scala/cats/Eval.scala) than IO, which is entirely stack-safe on one thread and no yields. I'm pretty sure the IOFiber runLoop in `cats-effect` is based off of the more basic `Eval` in cats, but don't quote me on that.

R.e. the current implementation: I thought I implemented a runLoop but I must not have. Oops!

Reply:
Yep, `Eval` is built off of a tail-recursive runLoop: https://github.com/typelevel/cats/blob/main/core/src/main/scala/cats/Eval.scala#L314